File: 0014_longest_common_prefix/trie/cpp/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <cassert>

class TrieNode {
public:
    std::unordered_map<char, TrieNode*> children;
    bool isEnd = false;
    int childCount = 0;
};

class Trie {
public:
    Trie() { root = new TrieNode(); }
    
    void insert(const std::string& word) {
        TrieNode* node = root;
        for (char c : word) {
            if (node->children.find(c) == node->children.end()) {
                node->children[c] = new TrieNode();
                node->childCount++;
            }
            node = node->children[c];
        }
        node->isEnd = true;
    }
    
    std::string findLongestPrefix() {
        TrieNode* node = root;
        std::string prefix = "";
        while (node->childCount == 1 && !node->isEnd) {
            auto it = node->children.begin();
            prefix += it->first;
            node = it->second;
        }
        return prefix;
    }

private:
    TrieNode* root;
};

class Solution {
public:
    std::string longestCommonPrefix(std::vector<std::string>& strs) {
        if (strs.empty()) return "";
        if (strs.size() == 1) return strs[0];
        
        Trie trie;
        for (const std::string& s : strs) {
            if (s.empty()) return "";
            trie.insert(s);
        }
        return trie.findLongestPrefix();
    }
};

void test() {
    Solution sol;
    std::vector<std::string> strs1 = {"flower", "flow", "flight"};
    assert(sol.longestCommonPrefix(strs1) == "fl");
    std::cout << "All tests passed!" << std::endl;
}

int main() {
    test();
    return 0;
}

File: 0014_longest_common_prefix/trie/python/solution.py
from typing import List, Dict

class TrieNode:
    def __init__(self):
        self.children: Dict[str, 'TrieNode'] = {}
        self.is_end = False
        self.child_count = 0

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str):
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
                node.child_count += 1
            node = node.children[char]
        node.is_end = True

    def find_longest_prefix(self) -> str:
        node = self.root
        prefix = ""
        while node.child_count == 1 and not node.is_end:
            char = list(node.children.keys())[0]
            prefix += char
            node = node.children[char]
        return prefix

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ""
        if len(strs) == 1:
            return strs[0]
            
        trie = Trie()
        for s in strs:
            if not s: return ""
            trie.insert(s)
            
        return trie.find_longest_prefix()
